### PR TITLE
Installed bcrypt to enable hashing. Implemented hashing.

### PR DESCRIPTION
**Pull request by aiharc1**
Researched and installed bcrypt to enable hashing. Implemented hashing.

Issue Number
#41 (password hashing)

Description
Researched possible methods on hashing/unhashing passwords effectively
Installed Bcrypt for hashing purposes (password)
Implemented password hashing when creating the user account with salt
Implemented unhashing at login page
-  user input plaintext password is compared with stored hashed password
-  upon comparing the two it will either match or not. 
-  when match, allow login. else, deny.

Testing
We will do formal test cases later. I have informally tested it with the following scenarios
- Correct user, correct password
- Invalid user, correct password
- Correct user, incorrect password
- Invalid user, invalid password
All of such tests passed. 

Possible Errors
bcrypt not being installed 

Images
The following image is the user object displaying the hashed password. 
If you'd like, the user/passoword for this user is minecraft/Spongebob1
image

### CODE REVIEW

**Comment by aiharc1** [comment]
Errors regaurding bcrypt being unidentified have been resolved. I installed in the incorrect location. It has been relocated to the correct location.

**Review by nguyenj21 — APPROVED** [review]
Bcrypt issue is fixed, so I will now work on writing test cases for the password hashing and username

File: small-talk/db/registerToDB.js [hunk 1]
```diff
@@ -1,14 +1,17 @@
...
 
+  const hashedpass = await bcrypt.hash(password, 10);
+
   const newUser = { // schema for holding the values into certain fields for database organization
     username: username,
-    password: password,
+    password: hashedpass,
     firstname: firstname,
     lastname: lastname,
     dob: date,
```
Comment (AlvaroH2001) [inline small-talk/db/registerToDB.js:17]: Doesn't have to be updated on this PR, but should consider adding in data validation of the user data, as it is it seems that the only validation so far is converting the data of birth to Unix.

**Review by AlvaroH2001 — APPROVED** [review]
Looks good so far and seems to get done what was expected, hashed passwords.

File: small-talk/db/registerToDB.js [hunk 2]
```diff
@@ -1,14 +1,17 @@
 "use server"
 const uri = "mongodb+srv://smt_root:pokemonwithguns@smalltalkcluster0.jo4jne6.mongodb.net/?retryWrites=true&w=majority"
 import Mongoboi from "./mongo"
+import bcrypt from "bcryptjs";
 export default async function handleRegister(username, password, firstname, lastname, dob){ // function that sends the credentials to the database
   const mongoboi = new Mongoboi(uri, "Users");
   const date = new Date(dob).getTime() / 1000; // return the inputted date as a Unix timestamp
   await mongoboi.connect();
 
+  const hashedpass = await bcrypt.hash(password, 10);
```
Comment (39otsu) [inline small-talk/db/registerToDB.js:10]: Great job salting. I would suggest to also find a way to generate a new salt for each user instance.

**Review by 39otsu — APPROVED** [review]
Excellent idea of employing the use of a "battle tested" module to handle an important aspect of authentication. Security. I also applaud the implementation of best practices.

I would also find a way to generate a random salt for each instance and store them along with the password as a separate field.

Refer to this post for more information
https://security.stackexchange.com/posts/17435/revisions

File: small-talk/db/registerToDB.js (hunk 1 above)
Comment (aiharc1) [reply small-talk/db/registerToDB.js:17]: Thank you for pointing this out. This is definitely something that should be addressed. We will create a new issue to handle this

File: small-talk/db/registerToDB.js (hunk 2 above)
Comment (aiharc1) [reply small-talk/db/registerToDB.js:10]: Thank you. We will look into how to implement this properly. Good idea

**Comment by aiharc1** [comment]
Since we have a total of 4 approvals I will go ahead and merge this PR. Thank you everyone

**Comment by aiharc1** [comment]
This is a diagram of how password hashing works
image

**Pull request merged**